virtual_ecosystem/models/abiotic/energy_balance.py

```diff
@@ -0,0 +1,649 @@
+    * absorbed_radiation: Absorbed radiation, [W m-2]
+    * specific_heat_air: Specific heat of air, [J mol-1 K-1]
+
+    TODO
+    * add latent heat flux from soil to atmosphere (-> VPD)
+    * check time integration
+    * set limits to temperature and VPD
+
+    Args:
+        data: Instance of data object
+        time_index: Time index
+        topsoil_layer_index: Index of top soil layer
+        abiotic_constants: Set of abiotic constants
+        abiotic_simple_constants: Set of abiotic constants
+        core_constants: Set of core constants
+
+    Returns:
+        air temperature, [C], canopy temperature, [C], vapour pressure [kPa], vapour
+        pressure deficit, [kPa]
+    """
+
+    output = {}
+
+    current_conductivities = calculate_current_conductivities(
+        data=data,
+        characteristic_dimension_leaf=np.repeat(
+            core_constants.characteristic_dimension_leaf,
+            data.grid.n_cells,
+        ),
+        von_karmans_constant=core_constants.von_karmans_constant,
+        abiotic_constants=abiotic_constants,
+    )
+
+    true_canopy_layers = data["leaf_area_index"][
+        data["leaf_area_index"]["layer_roles"] == "canopy"
+    ].dropna(dim="layers", how="all")
+
+    # Calculate vapour pressures
+    soil_saturated_vapour_pressure = calculate_saturation_vapour_pressure(
+        temperature=data["soil_temperature"][topsoil_layer_index],
+        factor1=abiotic_simple_constants.saturation_vapour_pressure_factor1,
+        factor2=abiotic_simple_constants.saturation_vapour_pressure_factor2,
+        factor3=abiotic_simple_constants.saturation_vapour_pressure_factor3,
+    )
+    soil_vapour_pressure = (
+        data["soil_moisture"][topsoil_layer_index] * soil_saturated_vapour_pressure
+    )
+    saturated_vapour_pressure_ref = calculate_saturation_vapour_pressure(
+        temperature=data["air_temperature_ref"].isel(time_index=time_index),
+        factor1=abiotic_simple_constants.saturation_vapour_pressure_factor1,
+        factor2=abiotic_simple_constants.saturation_vapour_pressure_factor2,
+        factor3=abiotic_simple_constants.saturation_vapour_pressure_factor3,
+    )
+
+    # Calculate conductivity from soil
+    conductivity_from_soil = (
+        data["soil_moisture"][topsoil_layer_index] * soil_saturated_vapour_pressure
+    )
+
+    # Factors from leaf and air temperature linearisation
+    a_A_list = []
+    for layer in range(1, len(true_canopy_layers) + 1):
+        a_A_layer = (
+            (
+                current_conductivities["conductivity_from_ref_height"][layer]
+                * data["air_temperature_ref"].isel(time_index=time_index)
+            )
+            + conductivity_from_soil * data["soil_temperature"][topsoil_layer_index]
+        ) / (
+            current_conductivities["conductivity_from_ref_height"][layer]
+            + conductivity_from_soil
+        )
+        a_A_list.append(a_A_layer)
+    a_A = np.vstack(a_A_list)
```

Comment:
Well, I initially did exactly that, extracting the true canopy layers and calculating the conductivities only for that. It turned out to be a bit of a nightmare to patch all values in the canopy and below together afterwards when there are the empty canopy layers in between, so I ended up carrying the DataArrays around. At that stage of development, I wrote the conductivities in the data object. That has changed since and might no longer require the full dimensions, so I'll look into it again.

Reply:
I replaced the loops with functions outside the long function, so hopefully that is clearer now.

@davidorme Regarding the mix of xarray and numpy array, I usually prefer to convert to numpy array at the start to avoid the mess with the dimension and coordinate names. Here, I didn't do that because some of the functions are from oder modules and expect DataArrays. I could go and change it there in the next iteration